geomstats/geometry/sasaki_metric.py

```diff
@@ -0,0 +1,238 @@
+
+
+def _gradient_descent(x_ini, grad, exp, lrate=0.1, max_iter=100, tol=1e-6):
+    """Apply a gradient descent until max_iter or a given tolerance is reached."""
+    L = len(x_ini)
+    x = x_ini
+    for _ in range(max_iter):
+        grad_x = grad(x)
+        grad_norm = gs.linalg.norm(grad_x)
+        if grad_norm < tol:
+            break
+        grad_x = -lrate * grad_x
+        for j in range(L):
+            x[j] = exp(grad_x[j], x[j])
+    return list(x)
+
+
+class SasakiMetric(RiemannianMetric):
+    """Implements of the Sasaki metric on the tangent bundle TM of a Riem. manifold M.
+
+    The Sasaki metric is characterized by the following three properties:
+     * the canonical projection of TM becomes a Riemannian submersion,
+     * parallel vector fields along curves are orthogonal to their fibres, and
+     * its restriction to any tangent space is Euclidean.
+
+    Geodesic computations are realized via a discrete formulation of the geodesic
+    equation on TM that involve geodesics, parallel translation, and the curvature
+    tensor on the base manifold M (see [1] for details).
+
+    Parameters
+    ----------
+    metric : RiemannianMetric
+        Metric of the base manifold of the tangent bundle.
+    n_s : int
+        Number of steps used in time-discrete geodesic computations.
+        Optional. Default : 3.
+
+    References
+    ----------
+    .. [1] Muralidharan, P., & Fletcher, P. T. (2012, June).
+    Sasaki metrics for analysis of longitudinal data on manifolds.
+    In 2012 IEEE conference on computer vision and pattern recognition (pp. 1027-1034).
+    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4270017/
+    """
+
+    def __init__(self, metric: RiemannianMetric, n_s=3):
+        self.metric = metric  # Riemannian metric of underlying space
+        self.n_s = n_s  # Number of discretization steps
+        shape = (2, gs.prod(metric.shape))
+
+        super(SasakiMetric, self).__init__(2 * metric.dim, shape=shape,
+                                           default_point_type='matrix')
+
+    def exp(self, tangent_vec, base_point, n_steps=None, **kwargs):
+        """Compute the Riemannian exponential of a point.
+
+        Exponential map at base_point of tangent_vec computed by
+        shooting a Sasaki geodesic using an Euler integration on TM.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., 2, M.dim]
+            Tangent vector in TTM at the base point in TM.
+        base_point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        n_steps : int
+            Number of discrete time steps.
+            Optional, default: self.n_s.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., 2, M.dim]
+            Point on the tangent bundle TM.
+        """
+        # unflatten
+        bs_pts = gs.reshape(base_point, (-1, 2) + self.metric.shape)
+        tngs = gs.reshape(tangent_vec, bs_pts.shape)
```

Comment:
Is reshaping such a big problem? As I understand it just changes attributes and does not touch the memory. For the moment I would want to keep the reshaping as it is, simply because I know it works for all relevant cases. Or is there another technical reason I am missing?

Reply:
You're write about reshaping. My alternative would be:

```
base_point = gs.broadcast_to(base_point, tangent_vec.shape)
```

In my opinion this solution is superior because:
* `gs.reshape` requires `self.metric.shape`. broadcast simply requires the arrays that are being manipulated.
* every time reshape is used we need to think "why am I reshaping?", "is this a proper reshaping?". whereas `gs.broadcast_arrays` clearly indicates the goal is to (properly) create arrays with the same shape.
* it is easier to manage the shape of the output (normally nothing needs to be done, whereas with reshape we need to update the shape).
* it is more compact.

A negative point is that the code will need to assume that we will handle one or several points, which sometimes may be tricky if we are not used to think in terms of vectorization.

Anyway, I'm OK with the solution you have.

Reply:
Ok, so we will keep the reshaping for the moment. And you (or anybody else) are/is free to refactor it in the future.